### PR TITLE
[ContactStructuralMechanicsApplication] Replace materials process

### PR DESCRIPTION
**Pull request by loumalouomega**
This process replaces the materials of a model during the simulation. For now it is in the ContactStructuralMechanicsApplication but can be moved to the Core or the StructuralMechanicsApplication due o its generality

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
why do you need reinitialization? initialize shoukd be done only once in principle, since there you get the rrference configuration to refer to.

for example i am not 100% sure that total lagrangian elements will work correctly if you teinitialize correctly in the moddle of calculations

**Comment by loumalouomega** [comment]
> why do you need reinitialization? initialize shoukd be done only once in principle, since there you get the rrference configuration to refer to.
> 
> for example i am not 100% sure that total lagrangian elements will work correctly if you teinitialize correctly in the moddle of calculations

In order to reload the CL. The Tl is not a problem, may be for UL. In any case the idea is more focus to reassign properties in conditions (to change friction for example). But decided to do it general

**Comment by loumalouomega** [comment]
I can add a flag

**Comment by loumalouomega** [comment]
Done

File: applications/ContactStructuralMechanicsApplication/python_scripts/replace_properties_process.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
+                material_settings["Parameters"]["materials_filename"].SetString(self.materials_filename)
+                KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
+
+                # Reinitilize elements and conditions
+                if self.reinitialize_entities:
+                    for elem in self.model_part.Elements:
+                        elem.Initialize(process_info)
+                    for cond in self.model_part.Conditions:
+                        cond.Initialize(process_info)
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/python_scripts/replace_properties_process.py:88]: +1 for using the new interface

File: applications/ContactStructuralMechanicsApplication/python_scripts/replace_properties_process.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            if settings["interval"][1].IsString():
+                if settings["interval"][1].GetString() == "End":
+                    settings["interval"][1].SetDouble(1e30) # = default_settings["interval"][1]
+                else:
+                    raise Exception("The second value of interval can be \"End\" or a number, interval currently:"+settings["interval"].PrettyPrintJsonString())
+
+        settings.ValidateAndAssignDefaults(default_settings)
+
+        # The model and model part
+        self.model = Model
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/python_scripts/replace_properties_process.py:53]: seems unused

**Review by philbucher — APPROVED** [review]
I don't know the usecase but to me this looks ok

I am curious, what is the usecase?

**Comment by loumalouomega** [comment]
> I don't know the usecase but to me this looks ok
> 
> I am curious, what is the usecase?

For replacing frictional laws (once they actually work...)

**Pull request merged**